csharp/Repl/RoslynScriptEngine.cs

```diff
@@ -0,0 +1,199 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Scripting;
+using Microsoft.CodeAnalysis.Scripting;
+using Microsoft.Spark.CSharp.Core;
+using Microsoft.Spark.CSharp.Sql;
+using Razorvine.Pickle;
+using Razorvine.Serpent;
+
+[assembly: InternalsVisibleTo("ReplTest")]
+namespace Microsoft.Spark.CSharp
+{
+    /// <summary>
+    /// Class that holds all objects shared in Roslyn runtime context.
+    /// </summary>
+    public class SparkCLRHost
+    {
+        public SparkContext sc;
+        public SqlContext sqlContext;
+    }
+
+    /// <summary>
+    /// A C# script engine based on Roslyn(https://github.com/dotnet/roslyn).
+    /// </summary>
+    public class RoslynScriptEngine : IScriptEngine
+    {
+        private int dumpSeq = 0;
+        private ScriptState<object> previousState;
+        private readonly SparkConf sparkConf;
+        private readonly SparkContext sc;
+        private readonly SparkCLRHost host;
+        private readonly ParseOptions options;
+
+        internal readonly string compilationDumpDirectory;
+
+        public RoslynScriptEngine(SparkContext sc)
+        {
+            this.sc = sc;
+            sparkConf = sc.GetConf();
```

Comment:
Right now ScriptEngine relies on SparkContext.addFile to ship the compiled dependencies across the cluster. 
But I agree that it's better to pass SparkContext and SqlContext in the constructor